README.md
# Battery Widget

<img src="images/BatteryWidget-gyu1.png" height="400" alt="Screenshot"/> <img src="images/BatteryWidget-gyu2.png" height="400" alt="Screenshot"/>

This project is based on Battery Widget by Richard Hyndman.  The original open source project became closed source from v1.7.3.


Graphical Battery Widget shows the exact battery level in 1x1 space, with shortcuts to Power-Summary/Display/Network/GPS/Wifi/BT Settings and 'Show Desktop' feature to shows your wallpaper/live-wallpaper in full screen.

**Note:**
This doesn't open, it is a Widget not an app.  Add to homescreen with

- Menu -> Add -> Widgets -> Battery Widget or
- Apps -> Widgets -> Battery Widget or
- Long press homescreen -> Widgets -> Battery Widget


## Installation
On Android 8.0 (Oreo) or above, if upon adding Battery Widget to the homescreen, the widget show 0% or stop updating, this is likely be related to the battery optimization feature introduced in Android 8.0, to fix this, disable battery optimization for Battery Widget by

- Battery Widget: App info -> Battery optimization -> Don't optimize

then

- Remove all Battery Widget instance from homescreen.
- Re-add Battery Widget to homescreen.


## Download
<a href="https://play.google.com/store/apps/details?id=gyu.geekyouup.android.widgets.battery">
  <img alt="Get it on Google Play" height="60" src="https://play.google.com/intl/en_us/badges/images/generic/en-play-badge.png"/>
</a>


## Acknowledgements
- Thanks to Richard Hyndman for [Battery Widget](http://code.google.com/p/batterywidget).


## License
```
Battery Widget - Android Home Screen Battery Widget
Copyright (c) 2014-2020 HairyRobot

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
```


Changelog.md
# Changelog

## v1.0.0
- Initial version is based on Battery Widget v1.7.2, with package name changed to gyu.geekyouup.android.widgets.battery.
- md5: dbd711f9230a26fb519d7b06bec6ef6d

## v1.1.0
- Migrated source to Android Studio.
- Bump minSdkVersion to 23 and targetSdkVersion to 25.
- Fixed 1x1 widget size bug on Android 10.
- Added dismiss dialog on touching outside dialog.
- md5: f253875a0f83a2f75eab8dc131d64e0d


src/gyu/geekyouup/android/widgets/battery/BatteryInfo.java
package gyu.geekyouup.android.widgets.battery;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.BatteryManager;
import android.util.Log;

public class BatteryInfo extends BroadcastReceiver {
	
	@Override
	public void onReceive(Context context, Intent intent) {
        try
        {
			String action = intent.getAction();
	        if (action.equals(Intent.ACTION_BATTERY_CHANGED)) {       		
	            SharedPreferences settings = context.getSharedPreferences(BatteryWidget.PREFS_NAME, Context.MODE_PRIVATE);
	            if(settings !=null)
	            {
	                int prevLevel = settings.getInt(BatteryWidget.KEY_LEVEL, -1);
	                int prevStatus = settings.getInt(BatteryWidget.KEY_CHARGING, -1);
	                
	                //for 1.6-
	                int currentLevel = intent.getIntExtra("level", 0);//BatteryManager.EXTRA_LEVEL for 2.0+
	                int currentStatus = intent.getIntExtra("status", BatteryManager.BATTERY_STATUS_UNKNOWN);//BatteryManager.EXTRA_STATUS for 2.0+

	                // Only update display if something changed.
	                if (prevLevel != currentLevel || prevStatus != currentStatus)
	                {
                        SharedPreferences.Editor editor = settings.edit();
                        editor.putInt(BatteryWidget.KEY_LEVEL, currentLevel);
	                    editor.putInt(BatteryWidget.KEY_CHARGING, currentStatus);
	                    int scale = intent.getIntExtra("scale", 100); //BatteryManager.EXTRA_SCALE for 2.0+
	                    editor.putInt(BatteryWidget.KEY_SCALE, scale);
	                    editor.commit();
	                    Intent forceUpIntent = new Intent(context, BatteryWidget.UpdateService.class);
	                    context.startService(forceUpIntent);
	               }
	            }
	        }
        }catch(Exception e){Log.e(BatteryWidget.LOG_TAG, "", e);}
	}
}


app/src/main/java/gyu/geekyouup/android/widgets/battery/cleaner/Cleaner.java
package gyu.geekyouup.android.widgets.battery.cleaner;

import android.app.Activity;
import android.app.WallpaperManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.widget.RelativeLayout;
import gyu.geekyouup.android.widgets.battery.R;

public class Cleaner extends Activity {
    private RelativeLayout mView;
	private WallpaperManager mWPM;
    
	/** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.cleaner);
    }
    
    @Override
    protected void onPause() {
    	super.onPause();
        
    	AlphaAnimation aa = new AlphaAnimation(1.0f, 0.0f);
        aa.setDuration(1000);
        mView.startAnimation(aa);
        finish();
    }
    
    @Override
    protected void onResume() {
    	super.onResume();

        mView = (RelativeLayout) findViewById(R.id.mainview);
        mView.setFocusableInTouchMode(false);

        if(android.os.Build.VERSION.SDK_INT>= Build.VERSION_CODES.ICE_CREAM_SANDWICH)
        {
            mView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE);
        }

        mWPM = WallpaperManager.getInstance(this);
    }
   
    float xPos = 0;
    float xWP = 0.5f;
    float yPos = 0;
    float yWP = 0.5f;
    @Override
    public boolean onTouchEvent(MotionEvent event) {
    	if(event.getAction() == MotionEvent.ACTION_DOWN)
    	{
    		xPos = event.getX();
    		yPos = event.getY();
    		//kill app if touched in bottom right corner
    		if(xPos > mView.getWidth()-50 && event.getY()>mView.getHeight()-50) finish();
    		else
    		{
    			mWPM.sendWallpaperCommand(mView.getWindowToken(), WallpaperManager.COMMAND_TAP, (int) xPos, (int) event.getY(), 0, null);
    		}
    	}else if(event.getAction() == MotionEvent.ACTION_MOVE)
    	{
    		xWP = xWP + (xPos-event.getX())/500;
    		xPos = event.getX();
    		if(xWP < 0) xWP = 0;
    		if(xWP > 1) xWP = 1;
    		
    		yWP = yWP + (yPos-event.getY())/500;
    		yPos = event.getY();
    		if(yWP < 0) yWP = 0;
    		if(yWP > 1) yWP = 1;
    		mWPM.setWallpaperOffsets(mView.getWindowToken(), xWP, yWP);
    	}
    	
    	if(android.os.Build.VERSION.SDK_INT>= Build.VERSION_CODES.ICE_CREAM_SANDWICH)
    	{
    		mView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE);
    		//Log.d("Cleaner", "Lights out");
    	}
    	
    	
    	return super.onTouchEvent(event);
    }
}

app/src/main/java/gyu/geekyouup/android/widgets/battery/TranslucentBlurActivity.java
package gyu.geekyouup.android.widgets.battery;

import java.lang.reflect.Method;
import java.util.List;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.location.LocationManager;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager; 
import android.view.View.OnClickListener;
import android.view.animation.AnimationUtils;
import android.widget.*;
import gyu.geekyouup.android.widgets.battery.cleaner.Cleaner;

/**
 * <h3>Fancy Blurred Background Activity</h3>
 */
public class TranslucentBlurActivity extends Activity implements OnClickListener {

	private Button mBtnDisplaySettings;
	private ToggleButton mBtnWifiSettings;
	private ToggleButton mBtnGPSSettings;
	private ToggleButton mBtnBluetoothSettings;
	private Button mBtnBatterySettings;
    private Button mBtnNetworkSettings;
    private Button mBtnShowDesktop;

    private Intent intentGPS;
    private Intent intentNetwork;

    private static final String LOG_TAG = "BatteryWidget";

    @Override
    protected void onCreate(Bundle icicle) {
        // Be sure to call the super class.
        super.onCreate(icicle);
        
        // Have the system blur any windows behind this one.
        if(!isLiveWallpaper(this) && BatteryWidget.SDK_VERSION<11) //no blur on honeycomb
        {
	        getWindow().setFlags(WindowManager.LayoutParams.FLAG_BLUR_BEHIND,
	                WindowManager.LayoutParams.FLAG_BLUR_BEHIND);
        }
		getWindow().setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL,
				WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL);
		getWindow().setFlags(WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH,
				WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH);

        setContentView(R.layout.settings);

        mBtnDisplaySettings = (Button) findViewById(R.id.displaySettings);
        mBtnDisplaySettings.setOnClickListener(this);
        
        mBtnWifiSettings = (ToggleButton) findViewById(R.id.wifiSettings);
        mBtnWifiSettings.setOnClickListener(this);
        
        WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(WIFI_SERVICE);
        mBtnWifiSettings.setChecked(wifiManager.isWifiEnabled());

        mBtnGPSSettings = (ToggleButton) findViewById(R.id.gpsSettings);
        mBtnGPSSettings.setOnClickListener(this);
        
        mBtnBluetoothSettings = (ToggleButton) findViewById(R.id.bluetoothSettings);
        mBtnBluetoothSettings.setOnClickListener(this);

        mBtnNetworkSettings = (Button) findViewById(R.id.networkSettings);
        mBtnNetworkSettings.setOnClickListener(this);

        mBtnShowDesktop = (Button) findViewById(R.id.cleaner);
        mBtnShowDesktop.setOnClickListener(this);

        if (BatteryWidget.SDK_VERSION > 3)
        {
	        mBtnBatterySettings = (Button) findViewById(R.id.batterySettings);
	        mBtnBatterySettings.setOnClickListener(this);
        }
        
        try
        {
        	LocationManager locManager = (LocationManager)getSystemService(LOCATION_SERVICE);
            boolean gpsOn = locManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
            mBtnGPSSettings.setChecked(gpsOn);
        	
        	int btStatus = Settings.Secure.getInt(getContentResolver(), android.provider.Settings.Secure.BLUETOOTH_ON);
        	mBtnBluetoothSettings.setChecked(btStatus == 1);
        }catch(Exception e)
        {
        	Log.d(LOG_TAG,"Failed to get Bluetooth status",e);
        }

        //disable the buttons that won't work
        intentGPS = new Intent("android.settings.LOCATION_SOURCE_SETTINGS");
        if(!isIntentAvailable(this, intentGPS)) mBtnGPSSettings.setEnabled(false);

        intentNetwork = new Intent(Settings.ACTION_DATA_ROAMING_SETTINGS);
        intentNetwork.addCategory("android.intent.category.DEFAULT");
        if(!isIntentAvailable(this, intentNetwork))
        {
            intentNetwork = new Intent(Settings.ACTION_WIRELESS_SETTINGS);
            if(!isIntentAvailable(this, intentNetwork))
            {
                mBtnNetworkSettings.setEnabled(false);
            }
        }

        ((LinearLayout) findViewById(R.id.settingsLayout)).startAnimation(AnimationUtils.loadAnimation(this, R.anim.launchanim));
    }
	
    protected void onPause() {
        super.onPause();
    }

    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }


	@Override
	public boolean onTouchEvent(MotionEvent event) {
		if (MotionEvent.ACTION_OUTSIDE == event.getAction())
		{
			finish();
			return true;
		}
		return super.onTouchEvent(event);
	}

	public void onClick(View v) {
		
		try
		{
			if(v==mBtnDisplaySettings)
			{
				Intent defineIntent2 = new Intent("com.android.settings.DISPLAY_SETTINGS");
				defineIntent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
				defineIntent2.addCategory("android.intent.category.DEFAULT");
				startActivity(defineIntent2);
				finish();
			}else if(v== mBtnWifiSettings)
			{
				WifiManager wifiManager = (WifiManager) getApplicationContext().getSystemService(WIFI_SERVICE);
				boolean enabled = wifiManager.isWifiEnabled();
				wifiManager.setWifiEnabled(!enabled);
				
				Toast.makeText(this, enabled?"Disabling Wifi":"Enabling Wifi", Toast.LENGTH_SHORT).show();
				finish();
			}else if(v==mBtnGPSSettings)
			{
				Intent defineIntent2 = new Intent("android.settings.LOCATION_SOURCE_SETTINGS");
				defineIntent2.addCategory("android.intent.category.DEFAULT");
				defineIntent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
				startActivity(defineIntent2);
				finish();
			}else if(v==mBtnBluetoothSettings)
			{
                Intent defineIntent2 = new Intent();
                if (BatteryWidget.SDK_VERSION >= 14) {
                    defineIntent2.setAction(Settings.ACTION_BLUETOOTH_SETTINGS);
                } else {
                    defineIntent2.setAction("android.settings.WIRELESS_SETTINGS");
                }

				defineIntent2.addCategory("android.intent.category.DEFAULT");
				defineIntent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
				startActivity(defineIntent2);
				finish();
			}else if(v==mBtnBatterySettings)
			{
				Intent defineIntent2 = new Intent("android.intent.action.POWER_USAGE_SUMMARY");
                defineIntent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
				startActivity(defineIntent2);
				finish();
			} else if(v==mBtnNetworkSettings)
            {
                startActivity(intentNetwork);
                finish();
            } else if(v==mBtnShowDesktop)
            {
                Intent defineIntent2 = new Intent(this,Cleaner.class);
                defineIntent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(defineIntent2);
                finish();
            }
		}catch(Exception e){Log.e(LOG_TAG,"Settings Error",e);}
	}
	
    public static boolean isLiveWallpaper(Context context)
    {
    	boolean liveWallpaper = false;
   		try
   		{
	   		if(BatteryWidget.SDK_VERSION>=7)
	   		{
	   			//this version of code needs Android 2.0, but we want to run on 1.5+, so use reflection
	   			//this version of code works on all vers of android
		   		Class classWallpaperManager = Class.forName("android.app.WallpaperManager");
		   		if(classWallpaperManager != null)
	        	{
	        		Method methodGetInstance = classWallpaperManager.getDeclaredMethod("getInstance", Context.class);
	        		Object objWallpaperManager = methodGetInstance.invoke(classWallpaperManager, context);
	        		
	        		Method methodGetWallpaperInfo = objWallpaperManager.getClass().getMethod("getWallpaperInfo", null);
	        		Object objWallPaperInfo = methodGetWallpaperInfo.invoke(objWallpaperManager, null);
	        		if(objWallPaperInfo!=null)
	        		{
	        			liveWallpaper=true;
	        		}
	        	}
	   		}
   		}catch(Throwable t){}
   		
   		return liveWallpaper;
    }

    public static boolean isIntentAvailable(Context context, Intent action) {
        try
        {
            final PackageManager packageManager = context.getPackageManager();
            List list = packageManager.queryIntentActivities(action,
                            PackageManager.MATCH_DEFAULT_ONLY);
            return list.size() > 0;
        }catch(Exception e)
        {
            Log.e(LOG_TAG, "Intent Checking Failed");
        }

        return true;
    }
}


app/src/main/java/gyu/geekyouup/android/widgets/battery/BatteryWidget.java
package gyu.geekyouup.android.widgets.battery;

import android.app.PendingIntent;
import android.app.Service;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.BatteryManager;
import android.os.IBinder;
import android.util.Log;
import android.view.View;
import android.widget.RemoteViews;

/**
 * Define a simple widget that shows a battery meter. To build
 * an update we spawn a background {@link Service} to perform the API queries.
 */
public class BatteryWidget extends AppWidgetProvider {
	
	public static final String LOG_TAG = BatteryWidget.class.getSimpleName();
	//public static final int SDK_VERSION = Integer.parseInt(android.os.Build.VERSION.SDK); //1.5 version
	public static final int SDK_VERSION = android.os.Build.VERSION.SDK_INT;
    public static final String KEY_SCALE = "KEY_SCALE";
	public static String PREFS_NAME="BATWIDG_PREFS";
	public static String KEY_LEVEL = "BATWIDG_LEVEL";
	public static String KEY_CHARGING = "BATWIDG_CHARGING";
	public static String KEY_VOLTAGE = "BATWIDG_VOLTAGE";
	
	@Override
	public void onEnabled(Context context) {
		super.onEnabled(context);
		BatteryWidget.clearSettings(context);
	}
	
	@Override
	public void onDisabled(Context context) {
		super.onDisabled(context);
		try
		{
			context.stopService(new Intent(context, UpdateService.class));//unregisterReceiver(mBI);
		}catch(Exception e){Log.d("BatteryWidget","Exception on disable: ",e);}
		BatteryWidget.clearSettings(context);
	}
	
	@Override
	public void onDeleted(Context context, int[] appWidgetIds) {
		super.onDeleted(context, appWidgetIds);
		try
		{
			context.stopService(new Intent(context, UpdateService.class));//if(mBI != null) context.unregisterReceiver(mBI);
		}catch(Exception e){Log.d("BatteryWidget","Exception on delete: ",e);}
	}
	
    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager,int[] appWidgetIds) {
        // To prevent any ANR timeouts, we perform the update in a service
    	context.startService(new Intent(context, UpdateService.class));
    }
    
    private static void clearSettings(Context context) {
        if (context != null)
        {
            SharedPreferences settings = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            Editor editor = settings.edit();
            editor.remove(KEY_LEVEL);
            editor.remove(KEY_CHARGING);
            editor.remove(KEY_SCALE);
            editor.commit();
        }
    }
    
    public static class UpdateService extends Service {
    	
    	BatteryInfo mBI = null;

    	@Override
        public void onStart(Intent intent, int startId) {
        	
			if(mBI == null)
	        {
	        	mBI = new BatteryInfo();
	        	IntentFilter mIntentFilter = new IntentFilter();
	            mIntentFilter.addAction(Intent.ACTION_BATTERY_CHANGED);
	            registerReceiver(mBI, mIntentFilter);
	        }
        	
        	// Build the widget update for today
        	RemoteViews updateViews = buildUpdate(this);
            if(updateViews != null)
            {
	            try
	            {
		            // Push update for this widget to the home screen
		            ComponentName thisWidget = new ComponentName(this, BatteryWidget.class);
		            if(thisWidget != null)
		            {
			            AppWidgetManager manager = AppWidgetManager.getInstance(this);
			            if(manager != null && updateViews != null)
			            {
			            	manager.updateAppWidget(thisWidget, updateViews);
			            }
		            }
		            
		            //stop the service, clear up memory, can't do this, need the Broadcast Receiver running
		            //stopSelf();
	            }catch(Exception e)
	            {
	            	Log.e("Widget", "Update Service Failed to Start", e);
	            }
            }
        }

    	@Override
    	public void onDestroy() {
    		super.onDestroy();
    		try{
    			if(mBI != null) unregisterReceiver(mBI);
    		}catch(Exception e)
    		{Log.e("Widget", "Failed to unregister", e);}
    	}
    	
        /**
         * Build a widget update to show the current Wiktionary
         * "Word of the day." Will block until the online API returns.
         */
        public RemoteViews buildUpdate(Context context) {           
            // Build an update that holds the updated widget contents
            RemoteViews updateViews = new RemoteViews(context.getPackageName(), R.layout.widget);
            try
            {
            	//Log.d("BatteryWidget","Updating Views");
	            int level = 0;
	            boolean charging = false;
	            SharedPreferences settings = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
	            if(settings !=null)
	            {
	            	level = settings.getInt(KEY_LEVEL, 0);
	            	
	            	//update level based on scale
	            	int scale = settings.getInt(KEY_SCALE, 100);
	            	if(scale != 100)
	            	{
			            if (scale <= 0) scale = 100;
			            level = (100 * level) / scale;
	            	}
	            	
		            charging = (settings.getInt(KEY_CHARGING, BatteryManager.BATTERY_STATUS_UNKNOWN)==BatteryManager.BATTERY_STATUS_CHARGING);
	            }
	            
	            if(level>20)
	            {
		            updateViews.setViewVisibility(R.id.bar100, level>80?View.VISIBLE:View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar80, level>60?View.VISIBLE:View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar60, level>40?View.VISIBLE:View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar40, level>20?View.VISIBLE:View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar20, View.VISIBLE);
		            updateViews.setImageViewResource(R.id.bar20, R.drawable.bar_green);
		            
		            updateViews.setViewVisibility(R.id.batterytext, View.VISIBLE);
	            }else
	            {
		            updateViews.setViewVisibility(R.id.bar100, View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar80, View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar60, View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar40, View.INVISIBLE);
		            updateViews.setViewVisibility(R.id.bar20, View.VISIBLE);
		            updateViews.setImageViewResource(R.id.bar20, R.drawable.bar_red);
		            updateViews.setViewVisibility(R.id.batterytext, View.VISIBLE);
	            }
	
	            updateViews.setViewVisibility(R.id.charging, charging?View.VISIBLE:View.INVISIBLE);
	            
	            String levelText = level==100?"100":level+"%"; //100% too wide
	            if(level == 0) levelText=" 0%";
	            updateViews.setTextViewText(R.id.batterytext, levelText);
	        }catch(Exception e)
	        {
	        	Log.e("BatteryWidget","Error Updating Views",e);
	        }
            
    		try
    		{
	            Intent defineIntent2 = new Intent(context,TranslucentBlurActivity.class);
	            PendingIntent pendingIntent2 = PendingIntent.getActivity(context,0 /* no requestCode */, defineIntent2, 0 /* no flags */);
	            updateViews.setOnClickPendingIntent(R.id.widget, pendingIntent2);
    		}catch(Exception e)
    		{
    			Log.e("BatteryWidget","Error Settings Intents",e);
    		}
            
            return updateViews;
        }
        
        @Override public IBinder onBind(Intent intent) {return null;}
    }

}
